packages/ui/src/lib/hooks/useEventsProvider.tsx

```diff
@@ -0,0 +1,175 @@
+	'stack-shapes': { operation: 'horizontal' | 'vertical' }
+	'flip-shapes': { operation: 'horizontal' | 'vertical' }
+	'distribute-shapes': { operation: 'horizontal' | 'vertical' }
+	'stretch-shapes': { operation: 'horizontal' | 'vertical' }
+	'reorder-shapes': {
+		operation: 'toBack' | 'toFront' | 'forward' | 'backward'
+	}
+	'delete-shapes': undefined
+	'select-all-shapes': undefined
+	'select-none-shapes': undefined
+	'rotate-ccw': undefined
+	'rotate-cw': undefined
+	'zoom-in': undefined
+	'zoom-out': undefined
+	'zoom-to-fit': undefined
+	'zoom-to-selection': undefined
+	'reset-zoom': undefined
+	'zoom-into-view': undefined
+	'zoom-to-content': undefined
+	'open-menu': { id: string }
+	'close-menu': { id: string }
+	'create-new-project': undefined
+	'save-project-to-file': undefined
+	'open-file': undefined
+	'select-tool': { id: string }
+	print: undefined
+	copy: undefined
+	paste: undefined
+	cut: undefined
+	'toggle-transparent': undefined
+	'toggle-snap-mode': undefined
+	'toggle-tool-lock': undefined
+	'toggle-grid-mode': undefined
+	'toggle-dark-mode': undefined
+	'toggle-focus-mode': undefined
+	'toggle-debug-mode': undefined
+	'exit-pen-mode': undefined
+	'stop-following': undefined
+}
+
+/** @public */
+export type TLUiEventHandler<T extends keyof TLUiEventMap = keyof TLUiEventMap> = (
+	source: string,
+	name: T,
+	data?: TLUiEventMap[T]
+) => void
+
+/** @internal */
+const defaultEventHandler: TLUiEventHandler = () => void null
+
+/** @internal */
+export const EventsContext = React.createContext({} as TLUiEventHandler)
+
+/** @public */
+export type EventsProviderProps = {
+	onEvent?: TLUiEventHandler
+	children: any
+}
+
+/** @public */
+export function EventsProvider({ onEvent, children }: EventsProviderProps) {
+	return (
+		<EventsContext.Provider value={onEvent ?? defaultEventHandler}>
+			{children}
+		</EventsContext.Provider>
+	)
+}
+
+/** @public */
+export function useEvents(): TLUiEventHandler<keyof TLUiEventMap> {
+	return React.useContext(EventsContext)
+}
+
+// export const REDACTED_VALUE = '[redacted]'
+
+// /**
+//  * Strip out tldraw sensitive data from the data value
+//  *
+//  * Note you shouldn't be just sending any old data to this and expecting it
+//  * to filter it out. However this is a safeguard to stop users being tracked
+//  * across sessions.
+//  *
+//  * @param item
+//  * @returns
+//  */
+// export function filterValue(item: string | number) {
+// 	if (typeof item === 'string') {
+// 		if (item.match(/^user:/)) {
+// 			return REDACTED_VALUE
+// 		} else {
+// 			return item
+// 		}
+// 	} else {
+// 		return item
+// 	}
+// }
+
+// export function filterObject(obj: Record<string, any>, fn: (item: any) => any) {
+// 	const out: Record<string, any> = {}
+// 	for (const [key, value] of Object.entries(obj)) {
+// 		if (filterValue(key) !== REDACTED_VALUE) {
+// 			out[key] = fn(value)
+// 		}
+// 	}
+// 	return out
+// }
+
+// export function filterSensitiveData(raw: any): any {
```

Comment:
This stuff is quite important as it makes sure we don't track uses by mistake.

Reply:
This seemed like a part of your branch where the events that were triggered as the result of changes to the store. We're no longer hooking into direct record changes, so it's probably more useful if we end up hooking into those sorts of changes externally (e.g. from the nextjs app).